File: src/providers/tavily.ts
/** Tavily — agent-friendly web search. POST https://api.tavily.com/search */

import {
  FetchedPage,
  MissingKeyError,
  NormalizedResult,
  Provider,
  ProviderError,
  httpRequest,
  safeGet,
} from "./base.js";

export const ENDPOINT = "https://api.tavily.com/search";
export const EXTRACT_ENDPOINT = "https://api.tavily.com/extract";

export interface TavilySearchOptions {
  maxResults?: number;
  searchDepth?: "basic" | "advanced";
  topic?: "general" | "news";
  days?: number;
  includeDomains?: string[];
  excludeDomains?: string[];
  country?: string;
  timeoutMs?: number;
}

export interface TavilyExtractOptions {
  extractDepth?: "basic" | "advanced";
  includeImages?: boolean;
  timeoutMs?: number;
}

export class TavilyProvider extends Provider {
  name = "tavily";
  schemaVersion = "tavily-v1-2026-04";
  apiKey: string | undefined;

  constructor(apiKey?: string) {
    super();
    this.apiKey = apiKey;
  }

  private ensureKey(): string {
    if (!this.apiKey) throw new MissingKeyError(this.name);
    return this.apiKey;
  }

  async search(query: string, opts: TavilySearchOptions = {}): Promise<NormalizedResult[]> {
    const body: Record<string, unknown> = {
      api_key: this.ensureKey(),
      query,
      max_results: opts.maxResults ?? 10,
      search_depth: opts.searchDepth ?? "basic",
    };
    if (opts.topic) body.topic = opts.topic;
    if (opts.days != null) body.days = opts.days;
    if (opts.includeDomains) body.include_domains = opts.includeDomains;
    if (opts.excludeDomains) body.exclude_domains = opts.excludeDomains;
    if (opts.country) body.country = opts.country;

    const response = await httpRequest(ENDPOINT, { method: "POST", body, timeoutMs: opts.timeoutMs });
    return this.normalize((response.json as Record<string, unknown>) ?? {});
  }

  normalize(payload: Record<string, unknown>): NormalizedResult[] {
    const results = (safeGet<unknown[]>(payload, ["results"], []) ?? []) as Array<Record<string, unknown>>;
    const out: NormalizedResult[] = [];
    for (const r of results) {
      const url = safeGet<string>(r, ["url"], "") ?? "";
      if (!url) continue;
      out.push(
        new NormalizedResult({
          url,
          title: safeGet<string>(r, ["title"], "") ?? "",
          snippet: safeGet<string>(r, ["content"], "") ?? "",
          score: toFloat(safeGet(r, ["score"])),
          publishedAt: safeGet<string>(r, ["published_date"]) ?? null,
          sourceKind: "web",
          provider: this.name,
          raw: r,
        }),
      );
    }
    return out;
  }
}

/**
 * Tavily Extract — used as a fetch fallback between Firecrawl and stdlib urllib.
 * Cheaper than Firecrawl, often handles Cloudflare-protected pages where
 * stdlib fails. Same TAVILY_API_KEY as the search provider.
 */
export class TavilyExtractProvider extends Provider {
  name = "tavily-extract";
  schemaVersion = "tavily-extract-v1-2026-04";
  apiKey: string | undefined;

  constructor(apiKey?: string) {
    super();
    this.apiKey = apiKey;
  }

  private ensureKey(): string {
    if (!this.apiKey) throw new MissingKeyError(this.name);
    return this.apiKey;
  }

  async extract(url: string, opts: TavilyExtractOptions = {}): Promise<FetchedPage> {
    const body: Record<string, unknown> = {
      api_key: this.ensureKey(),
      urls: [url],
      extract_depth: opts.extractDepth ?? "basic",
      include_images: opts.includeImages ?? false,
    };
    const response = await httpRequest(EXTRACT_ENDPOINT, {
      method: "POST",
      body,
      timeoutMs: opts.timeoutMs ?? 45_000,
    });
    const payload = (response.json as Record<string, unknown>) ?? {};
    const results = (safeGet<unknown[]>(payload, ["results"], []) ?? []) as Array<Record<string, unknown>>;
    const failed = (safeGet<unknown[]>(payload, ["failed_results"], []) ?? []) as Array<Record<string, unknown>>;
    if (results.length === 0) {
      const reason = failed.length > 0 ? String(safeGet(failed[0], ["error"], "no content")) : "no content";
      throw new ProviderError(`tavily-extract: ${reason} for ${url}`);
    }
    const r = results[0]!;
    const raw = String(safeGet<string>(r, ["raw_content"], "") ?? "");
    const sourceUrl = String(safeGet<string>(r, ["url"], url) ?? url);
    if (!raw) throw new ProviderError(`tavily-extract: empty raw_content for ${url}`);
    return new FetchedPage({
      url: sourceUrl,
      title: deriveTitle(raw, sourceUrl),
      markdown: raw.slice(0, 200_000),
      provider: this.name,
      status: "degraded",
      metadata: { sourceURL: sourceUrl, extractDepth: opts.extractDepth ?? "basic" },
    });
  }
}

function deriveTitle(content: string, fallback: string): string {
  // Tavily returns plain text + extracted links. Try to grab the first
  // non-empty line; fall back to the URL.
  const lines = content.split("\n").map((l) => l.trim()).filter(Boolean);
  return lines[0]?.slice(0, 200) || fallback;
}

function toFloat(v: unknown): number | null {
  if (v == null) return null;
  const n = Number(v);
  return Number.isFinite(n) ? n : null;
}

File: cli/src/wsc/providers/tavily.py
"""Tavily — agent-friendly web search (https://docs.tavily.com/).

Endpoint: ``POST https://api.tavily.com/search``.
Auth: API key in body (``api_key`` field). The newer header form
(``Authorization: Bearer ...``) is also accepted; v0.1 uses the body form
for maximum compatibility with older accounts.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from wsc.providers.base import (
    NormalizedResult,
    Provider,
    http_request,
    MissingKeyError,
    safe_get,
)

ENDPOINT = "https://api.tavily.com/search"


class TavilyProvider(Provider):
    name = "tavily"
    schema_version = "tavily-v1-2026-04"

    def __init__(self, api_key: Optional[str] = None) -> None:
        self.api_key = api_key

    def _ensure_key(self) -> str:
        if not self.api_key:
            raise MissingKeyError(self.name)
        return self.api_key

    def search(
        self,
        query: str,
        *,
        max_results: int = 10,
        search_depth: str = "basic",  # "basic" | "advanced"
        topic: Optional[str] = None,  # "general" | "news"
        days: Optional[int] = None,   # only used when topic == "news"
        include_domains: Optional[List[str]] = None,
        exclude_domains: Optional[List[str]] = None,
        country: Optional[str] = None,
        timeout: float = 30.0,
    ) -> List[NormalizedResult]:
        body: Dict[str, Any] = {
            "api_key": self._ensure_key(),
            "query": query,
            "max_results": int(max_results),
            "search_depth": search_depth,
        }
        if topic:
            body["topic"] = topic
        if days is not None:
            body["days"] = int(days)
        if include_domains:
            body["include_domains"] = include_domains
        if exclude_domains:
            body["exclude_domains"] = exclude_domains
        if country:
            body["country"] = country

        response = http_request(ENDPOINT, method="POST", body=body, timeout=timeout)
        return self.normalize(response.get("json") or {})

    def normalize(self, payload: Dict[str, Any]) -> List[NormalizedResult]:
        results = safe_get(payload, "results", default=[]) or []
        out: List[NormalizedResult] = []
        for r in results:
            url = safe_get(r, "url", default="")
            if not url:
                continue
            out.append(
                NormalizedResult(
                    url=url,
                    title=safe_get(r, "title", default="") or "",
                    snippet=safe_get(r, "content", default="") or "",
                    score=_to_float(safe_get(r, "score")),
                    published_at=safe_get(r, "published_date"),
                    source_kind="web",
                    provider=self.name,
                    raw=r,
                )
            )
        return out


def _to_float(v: Any) -> Optional[float]:
    if v is None:
        return None
    try:
        return float(v)
    except (TypeError, ValueError):
        return None
